Cleanup sv utils

SHuang-Broad opened the pull request:
Periodic housekeeping.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
I think that this method, with some refactoring, could belong to `IOUtils`.

SHuang-Broad replied:
Can you elaborate, specifically what refactoring is needed?

magicDGS replied:
I think that it would work by using the following code (but I haven't test it):

```java
// this should take care of GCS
final Path dir = IOUtils.getPath(Utils.nonNull(pathString));
if (Files.exists(dir)) {
    throw new IOException("Directory to be created already exists: " + dir.toUri());
}
// maybe some exceptions should be catch to return false (I am not sure about the usage of the boolean)
Files.createDirectory(dir);
return Files.isDirectory(dir) & Files.isWritable(dir);
```

In that case, it belongs to `IOUtils`, because it can be re-used easily...

SHuang-Broad replied:
I've done mostly what you suggested, can you take a look again?

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
I think that the `java.nio.Path` should be created with `IOUtils` here, and outside to re-use afterwards. Then the method could be simplify, I guess...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
Cannot the code using this be changed to `Files.write(Path, Iterable<? extends CharSequence>, OpenOption...)`. If the `Path` is obtained with `IOUtils`, then it should be fine if it is in a bucket too...

magicDGS replied:
I still thing that the usage of this method can be substituted by `Files.write(IOUtils.getPath(fileName), linesToWrite)`; there are also utils already implemented to convert between iterable, iterators and stream (see `org.broadinstitute.hellbender.utils.Utils`).

SHuang-Broad replied:
@magicDGS Thanks! I've replaced this method with the one you suggested and it is working well.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
Maybe this could be moved to `SparkUtils`... may be useful outside the SV code...

SHuang-Broad replied:
This method is not really tested well I am hesitated to move it there. Beside I am not the author.
What usage do you have in mind?

magicDGS replied:
I have an usage case where I need a tool iterating over kmers in the reference: although still in development, I have a reference walker by kmers, and at some point it would be nice to have also a spark-based reference walker.

Anyway, it was just a suggestion. For me, it is just simpler to look for useful methods in the utils classes related to the stuff I'm looking for (e.g., `CigarUtils` for cigar, `SparkUtils`for RDD related stuff, etc.

SHuang-Broad replied:
@magicDGS in that case (considering your tool is still under dev), I intend to keep it here and get more test coverage when you think you are ready. Is that OK?

magicDGS replied:
It's perfectly fine, I'm just worried that I forget that this is here and implement my own method. Usually I look at the `SparkUtils` code for this kind of things, and I do not use to look at the SV code at all except in some PRs with functionality that I think that I can use as a third-party developer...

SHuang-Broad replied:
@magicDGS . In that case, I'll try to add tests and move it to SparkUtils. In the meantime, if you find other methods in `SVUtils` and `SvCigarUtils` that are useful to you, let me know.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SvCigarUtils.java:
```diff
@@ -377,4 +377,11 @@ public static int computeAssociatedDistOnRead(final Cigar cigarAlong5To3DirOfRea
         return readWalkDist;
     }
+
+    public static int getTotalAlignmentLength(final Cigar cigar ) {
```
Maybe is it useful in `CigarUtils` to re-use outside SV code? I do not see specificity with SV here...

vruano replied:
I agree with @magicDGS, you can call it something like ```countAlignedBases```.

SHuang-Broad replied:
done

SHuang-Broad commented:
@vruano would you mind taking a look?

SHuang-Broad commented:
@magicDGS , I'll address your comments in the next commit. Thanks!

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
This method is not just public visible for testing; is referenced from the main code from a different package so please remove this annotation.

SHuang-Broad replied:
removed

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
Check arguments are not null or otherwise invalid.

vruano replied:
I think that getRefRDD and processRefRDD are doing a very specific kind of work on the reference genome and so I don't think they would qualify as utilities ... they seem to be only used by FindBadGenomeKmers, so I would suggest that these methods are moved to that class.

SHuang-Broad replied:
checked.

SHuang-Broad replied:
`getRefRDD` was in `SVUtils` before so I moved it here thinking it is providing a special type of utility related to reference. So I'll leave it here.
Moved processRefRDD back to `FindBadGenomeKmers`

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
Check arguments are not null or otherwise invalid.

SHuang-Broad replied:
done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
I guess the name is again not informative enough...

SHuang-Broad replied:
renamed

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVUtils.java:
```diff
@@ -165,6 +39,13 @@ public static int iteratorSize( final LongIterator itr ) {
     }
 
+    public static Collection<SVKmer> uniquify(final Collection<SVKmer> coll1, final Collection<SVKmer> coll2) {
+        final HopscotchSet<SVKmer> kmers = new HopscotchSet<>(coll1.size() + coll2.size());
```
Without looking at the uses... I guess you again simply added coll2 elements into coll1 and return coll1.

vruano replied:
Since is public you must check arguments are legal (non null).

SHuang-Broad replied:
done checking. If second is added to 1st, the 1st is modified, unlike what it is doing now (copying from the 1st).

vruano replied:
My understanding is that you don't want to make the change because you think it will break it? I don't think it would but I guess I have to find the reference supporting my claim.

SHuang-Broad replied:
Not necessarily breaking it, but that would change the behavior (and of course the signature) of the function.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVUtils.java:
```diff
@@ -207,6 +88,11 @@ private void advance() {
+    // =================================================================================================================
+
+    //Workaround for seed 14695981039346656037 that doesn't fit in a signed long
+    private static final long FNV64_DEFAULT_SEED = new BigInteger("14695981039346656037").longValue();
```
Curious what is the need to use a "problematic" seed number.

SHuang-Broad replied:
not sure. That's a question for TedS.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
Does any subtype of reference source every going to return a null without failing first?

SHuang-Broad replied:
not sure. This check was there before the refactoring.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java:
```diff
@@ -161,7 +162,7 @@ protected void runTool( final JavaSparkContext ctx ) {
 
             final SAMFileHeader cleanHeader = new SAMFileHeader(header.getSequenceDictionary());
-            final List<String> refNames = AlignedAssemblyOrExcuse.getRefNames(header);
+            final List<String> refNames = SVReferenceUtils.getRefNames(header);
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/AlignedAssemblyOrExcuse.java:
```diff
@@ -355,4 +241,97 @@ public AlignedAssemblyOrExcuse read( final Kryo kryo, final Input input, final C
         }
     }
+
+    public static Stream<SAMRecord> toSAMStreamForOneContig( final SAMFileHeader header, final List<String> refNames,
```
done in a way mostly as suggested, except that kept the required argument of `refNames` for performance concerns (so we don't go through the streaming of getting the ref contig names for every read/assembly-contig).

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/AlignedAssemblyOrExcuse.java:
```diff
@@ -355,4 +241,97 @@ public AlignedAssemblyOrExcuse read( final Kryo kryo, final Input input, final C
+                                          final SAMFileHeader header,
+                                          final List<SVInterval> intervals,
+                                          final List<AlignedAssemblyOrExcuse> intervalDispositions ) {
+        final Map<Integer, AlignedAssemblyOrExcuse> resultsMap = new HashMap<>();
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/AlignedAssemblyOrExcuse.java:
```diff
@@ -355,4 +241,97 @@ public AlignedAssemblyOrExcuse read( final Kryo kryo, final Input input, final C
+                      new OutputStreamWriter(new BufferedOutputStream(BucketUtils.createFile(intervalFile))) ) {
+            final List<SAMSequenceRecord> contigs = header.getSequenceDictionary().getSequences();
+            final int nIntervals = intervals.size();
+            for ( int intervalId = 0; intervalId != nIntervals; ++intervalId ) {
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/AlignedAssemblyOrExcuse.java:
```diff
@@ -355,4 +241,97 @@ public AlignedAssemblyOrExcuse read( final Kryo kryo, final Input input, final C
+            final List<SAMSequenceRecord> contigs = header.getSequenceDictionary().getSequences();
+            final int nIntervals = intervals.size();
+            for ( int intervalId = 0; intervalId != nIntervals; ++intervalId ) {
+                final SVInterval interval = intervals.get(intervalId);
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/AlignedAssemblyOrExcuse.java:
```diff
@@ -355,4 +241,97 @@ public AlignedAssemblyOrExcuse read( final Kryo kryo, final Input input, final C
+                                      final SAMFileHeader header,
+                                      final SAMFileHeader.SortOrder assemblyAlnSortOrder,
+                                      final String outputAssembliesFile ) {
+        final SAMFileHeader cleanHeader = new SAMFileHeader(header.getSequenceDictionary());
```
The header is requested several lines down.

SHuang-Broad replied:
You are right. Thanks for catch that!

SHuang-Broad replied:
also changed the order of parameters so the 1st parameter is the file name as you suggested somewhere else.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/AlignedAssemblyOrExcuse.java:
```diff
@@ -355,4 +241,97 @@ public AlignedAssemblyOrExcuse read( final Kryo kryo, final Input input, final C
+                        disposition + "\n");
+            }
+        } catch ( final IOException ioe ) {
+            throw new GATKException("Can't write intervals file " + intervalFile, ioe);
```
No. This output file is for debugging or tuning only so user doesn't directly control that. So no action.

SHuang-Broad replied:
did what you suggested on this round of cleanup.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSpark.java:
```diff
@@ -141,13 +140,16 @@ public static AssembledEvidenceResults gatherEvidenceAndWriteContigSamFile(
+        // write alignments of the assembled contigs
+        AlignedAssemblyOrExcuse.writeAssemblySAMFile(alignedAssemblyOrExcuseList, header, params.assembliesSortOrder, outputAssemblyAlignments);
         log("Wrote SAM file of aligned contigs.", toolLogger);
+//        // write the output file
```
:scream_cat: removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/KmerCounter.java:
```diff
@@ -25,7 +26,7 @@ public KmerCounter( final int kSize, final int kmersPerPartitionGuess,
     }
 
-    public Iterator<KmerAndCount> apply( final Iterator<GATKRead> readItr ) {
+    public Iterator<KmerAndCount> apply(final Iterator<GATKRead> readItr ) {
```
added, as that seems to be TedS' style, who is the author as well.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSpark.java:
```diff
@@ -451,8 +453,8 @@ public EvidenceScanResults(final ReadMetadata readMetadata,
                         new ReadsForQNamesFinder(broadcastQNamesMultiMap.value(), nIntervals,
                                 includeMappingLocation, readItr, filter).iterator(), false)
                 .combineByKey(x -> x,
```
I refactored/moved the `combineLists` from FindBreakpointEvidence to SVFastqUtils, so would rather let Ted handle this as he has many acrobatic tricks on how to avoid shuffles, and I refrain from messing with that.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFastqUtils.java:
```diff
@@ -48,6 +48,15 @@
 
     private static final Pattern SA_TAG_ALN_INTERVAL_SEPARATOR_PATTERN = Pattern.compile(HEADER_FIELD_LIST_SEPARATOR_STR);
 
+    /** Concatenate two lists. */
```
Again this is a simple move from another class, so would not mess with it right now.
Also, like you said this is not necessarily for Spark related jobs. So not messing with it yet in this PR.

SHuang-Broad replied:
after 2nd thought, refactored into SVUtils.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
What would you suggest for the name?
I am creating a dir to write some (SAM) files in a prototyping tool.

vruano replied:
```createDirectory```? I just don't why would you create a directory if you are not going to write into it, you or something else later on so that part of the name is rather pointless.... That said it seems this method would better belong to BucketUtils pairing with ```createFile``` as I say below.

SHuang-Broad replied:
renamed.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
done in its new place

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/ExtractOriginalAlignmentRecordsByNameSparkIntegrationTest.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
moved check over there

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
added suffix "*Path"

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
added suffix "*Path"

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
added suffix "*Path"

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
local enough and used only once, so replace with commented number.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
Not really my code initially. So I'll leave it to the next round of clean up.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
available only to BufferedWriter but not Writer. BufferedWriter has `write(String, int, int )` but no `write(String)`, so no action considering trade-off.

vruano replied:
Nothing is stopping you from change the type of ```writer``` to ```BufferedWriter```, but fair enough, is not your code.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
extension check pushed up.

SHuang-Broad replied:
truth be told, the code that I moved and refactored from does this file creation. 
And I guess the `SAMFileWriterFactory` version doesn't take hdfs path, which this BucketUtils version does allow.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
changed to 
```java
final Writer writer =
   new BufferedWriter(new OutputStreamWriter(BucketUtils.createFile(fileName))
```
which doesn't have `newLine()` but does have `write(String)`.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
checking done

SHuang-Broad replied:
done reordering parameters

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
changed to consistent behavior of returning boolean

SHuang-Broad replied:
lied. Changed to throwing IOException and no return.

SHuang-Broad commented:
@vruano and @magicDGS  I made most of the changes you requested, with the following exceptions:

* methods that were complex enough and I am not the original author&mdash;I didn't change most of them except adding some arg checking
* `SVFileUtils.createDirectory()` not sure if the change is what you would like to see (most likely still needs some work) and it can be moved to `BucketUtils` (I need to add tests for it before that)
* `SVFileUtils.writeSAMFile()` maybe it is worth to move it to some engine utils, but I need to talk to some Engine guy first.

SHuang-Broad commented:
Regarding writing the SAM file, I now remember why it was created instead of using existing ones:  the SV pipeline needs to be able to write to HDFS, hence cannot rely on a `File` interface&mdash;which the `ReadUtils.createCommonSAMWriter(...)` exposes. 
So the current hand rolled version in `SVFileUtils.getSAMFileWriter()` calls into `BucketUtils.createFile(...)` for that HDFS compatibility, and then makes use of the `SAMFileWriterFactory.makeBAMWriter(final SAMFileHeader header, final boolean presorted, final OutputStream stream)`, unlike `ReadUtils.createCommonSAMWriter(...)` which calls into `SAMFileWriterFactory.makeSAMOrBAMWriter(final SAMFileHeader header, final boolean presorted, final Path outputPath)`.

So in short: HDFS compatibility.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBadGenomicKmersSpark.java:
```diff
@@ -77,11 +84,63 @@ protected void runTool( final JavaSparkContext ctx ) {
+                .mapPartitions(seqItr -> {
+                    final HopscotchMap<SVKmer, Integer, KmerAndCount> kmerCounts = new HopscotchMap<>(hashSize);
+                    while ( seqItr.hasNext() ) {
+                        final byte[] seq = seqItr.next();
```
What about:
```
Utils.stream(seqItr)
       .flatMap(seq -> SVDUSTFilteredKmerizer.canonicalStream(seq, kSize, maxDUSTScore, new SVKmerLong())
       .forEach(kmer -> {...});
```

If one defines a ```Hopscotch.bumpAll(Hopscotch<K, V, T> other)``` that would add the counts from the other hop-scotch then one could use the ```Stream.reduce(Unit, Add, Combine)``` method instead than the ```forEach```.

SHuang-Broad replied:
not going to change implementations in this PR, since this method is not being refactored into Utils any more. Trying to limit what this PR does.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBadGenomicKmersSpark.java:
```diff
@@ -77,11 +84,63 @@ protected void runTool( final JavaSparkContext ctx ) {
+                })
+                .mapToPair(entry -> new Tuple2<>(entry.getKey(), entry.getValue()))
+                .partitionBy(new HashPartitioner(nPartitions))
+                .mapPartitions(pairItr -> {
```
You could do something similar here.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/KmerAndCount.java:
```diff
@@ -18,7 +18,7 @@
     public KmerAndCount( final SVKmerLong kmer ) { this(kmer,1); }
 
-    KmerAndCount( final SVKmerLong kmer, final int count ) {
+    public KmerAndCount( final SVKmerLong kmer, final int count ) {
```
Since is public, must it check that the ```kmer``` is not null and ```count``` is 0 or positive?

SHuang-Broad replied:
added a check for non-negative number.
But for the null check, I don't quite understand yet what the benefit is for this specific one. It delegates to the parent class's ctor, which would immediately throw an NPE if it null. An explicit null check would pretty much to the same thing, right?

vruano replied:
As per GATK3  GATK4 policy NPE are not acceptable; you can argue that with @droazen but I think is a lost cause, right David?

SHuang-Broad replied:
added null check

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
```diff
@@ -108,35 +98,37 @@ private static SAMFileWriter getSAMFileWriter(final String outputName, final SAM
         }
         catch ( final IOException ioe ) {
-            throw new GATKException("Unable to read kmers from "+kmersFile, ioe);
+            throw new GATKException("Unable to read kmers from "+kmersFilePath, ioe);
```
Spaces around "+", just bother if you are going to change this file for any other reason , not for this alone.

SHuang-Broad replied:
done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
```diff
@@ -108,35 +98,37 @@ private static SAMFileWriter getSAMFileWriter(final String outputName, final SAM
         }
         catch ( final IOException ioe ) {
-            throw new GATKException("Unable to write kmers to "+kmersFile, ioe);
+            throw new GATKException("Unable to write kmers to "+kmersFilePath, ioe);
```
same with the "+"

SHuang-Broad replied:
done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
```diff
@@ -147,7 +139,7 @@ private static SAMFileWriter getSAMFileWriter(final String outputName, final SAM
                 final String[] tokens = line.split("\t");
                 if ( tokens.length != 3 ) {
-                    throw new GATKException("Interval file "+intervalsFile+" line "+
+                    throw new GATKException("Interval file "+intervalsFilePath+" line "+
```
Same.

SHuang-Broad replied:
done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
```diff
@@ -158,27 +150,27 @@ private static SAMFileWriter getSAMFileWriter(final String outputName, final SAM
                     intervals.add(new SVInterval(contigId, start, end));
                 }
                 catch ( final Exception e ) {
-                    throw new GATKException("Unable to parse interval file "+intervalsFile+" line "+lineNo+": "+line, e);
```
Won't repeat any more.

SHuang-Broad replied:
done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVReferenceUtils.java:
```diff
@@ -39,16 +28,18 @@
-        if ( dict == null ) throw new GATKException("No reference dictionary available");
-
-        final int effectiveRecLen = ref_record_len - kSize + 1;
+    public static JavaRDD<byte[]> getReferenceBasesRDD(final JavaSparkContext ctx,
```
I still think this is not general enough to be a utility.... where do you think is it going to use? perhaps it could be made a bit more general if instead of returning chunks of arbitrary length it returns well defined kmer so you would hide the intermediary chanck size.

SHuang-Broad replied:
@magicDGS thinks he would have a use for that. 
I think this could be useful as well. 
I'll what improvements (test, representation) I can bring to this.

vruano replied:
What most using code would find useful is ```JavaRDD<byte[]>``` of all the khmers of a given length rather than a collection of chunks of bases or different sizes.... I guess the reason for returning the later is because to create a JavaRDD it seems that you need to explicitly create a List of all the its elements (is not possible to provide a stream instead?) (so ```ctx.parallelize(kmerList)```... this obviously not practical for a large genome like the human genome. 

I think the best way to approach this by changing the actual RDD that is returned. 
Instead of returning the RDD of the DNA chunks as we do now we return a ```flatMap``` on it that would transform those chunks into the actual kmers so these are only instantiated when they are needed... then the parameters of this method except for the reference source and the kmerSize can be hidden given them reasonable defaults.

Perhaps instead of byte[] you could return SVKmers objects.

I leave it up to you but the chuck-of-DNA RDD return looks a bit nasty and complex as a return type and given its specific about k-1 overlapping ... it is obvious what the intention for it so why don't simply return a RDD to the kmers that is what people really want.

SHuang-Broad replied:
I've added a todo for this method, but don't think I have time for it this week or next.
Also cleaned up the documentation a bit.

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemAlignmentUtils.java:
```diff
@@ -130,4 +131,22 @@ public static String asTag( final BwaMemAlignment alignment, final List<String>
+    public static Stream<SAMRecord> toSAMStreamForRead(final String readName, final byte[] contigSequence,
+                                                       final List<BwaMemAlignment> alignments,
+                                                       final SAMFileHeader header, final List<String> refNames) {
+        if ( alignments.isEmpty() ) return Stream.empty();
```
Check arguments.

SHuang-Broad replied:
done

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
I still cannot understand why a Hadoop path should be treated different than another kind of path. I believe that the `IOUtils.getPath` will get properly the Hadoop path, and then `Files.createDirectory` should get the `HadoopFileSystemProvider` for the HDFS, which in turn uses the mkdirs that you are using here.

Sorry if I am missing something....

SHuang-Broad replied:
done what's suggested here, moved to IOUtils and tested

vruano replied:
@magicDGS, in theory you are right but in practice there seems to be some issues (see https://github.com/broadinstitute/gatk/issues/3569 https://github.com/broadinstitute/barclay/issues/106)
Perhaps now we could drop the handoop but as far as this PR is concern it should follow the current approach.

magicDGS replied:
Actually, the problem is mainly with the GCS provider as far as I know (If I am wrong, maybe @lbergelson or @droazen can chime in for correct me).

The discussion in barclay is pointing to the method in `IOUtils`, as the hack for getting correctly the paths, which is working perfectly fine for hadoop (I have been using it in my own code properly).

In the case of this code, if the path is get by `IOUtils.getPath` the providers are loaded on demand (including the special case of GCS, which is handled with `BucketUtils`). Thus, using `Files` is possible once the providers are loaded. Again, as far as I know, the __jsr203-hadoop__ provider is working perfectly fine with `Files` (tested in my own toolkit) and that is the only special case handled here. I still do not fully understand why that's the case, while the GCS code is handled by `Files` in the implementation pointed by your comment. In principle, I think that most of the problems with providers are found with the GCS loading code...but again, @lbergelson or @droazen can correct me if I am missing something...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
The `ReadUtils` class will have support for HDFS and other `java.nio.Path` representation here: https://github.com/broadinstitute/gatk/pull/2558

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/SvDiscoverFromLocalAssemblyContigAlignmentsSpark.java:
```diff
@@ -19,6 +19,7 @@
 import org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary;
 import org.broadinstitute.hellbender.engine.spark.GATKSparkTool;
 import org.broadinstitute.hellbender.exceptions.GATKException;
```
removed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBadGenomicKmersSpark.java:
```diff
@@ -32,6 +36,9 @@
         programGroup = StructuralVariationSparkProgramGroup.class)
 @BetaFeature
 public final class FindBadGenomicKmersSpark extends GATKSparkTool {
+    public static final int REF_RECORD_LEN = 10000;
```
I agree, but since this is the code already in place in master, and after I moved back the intended-to-be-refactored out method back here, it is not used by any other code, I'll say let's keep it as is.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBadGenomicKmersSpark.java:
```diff
@@ -32,6 +36,9 @@
+    // assuming we have ~1Gb/core, we can process ~1M kmers per partition
+    public static final int REF_RECORDS_PER_PARTITION = 1024*1024 / REF_RECORD_LEN;
     private static final long serialVersionUID = 1L;
     @VisibleForTesting static final int MAX_KMER_FREQ = 3;
```
changed

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBadGenomicKmersSpark.java:
```diff
@@ -77,11 +84,63 @@ protected void runTool( final JavaSparkContext ctx ) {
+     * Kmerize, mapping to a pair <kmer,1>, reduce by summing values by key, filter out <kmer,N> where
+     * N <= MAX_KMER_FREQ, and collect the high frequency kmers back in the driver.
+     */
+    public static List<SVKmer> collectUbiquitousKmersInReference(final int kSize,
```
correct. Changed.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFileUtils.java:
```diff
@@ -15,42 +15,31 @@
-            }
-        } catch ( final IOException ioe ) {
-            throw new GATKException("Can't write "+fileName, ioe);
+    public static void createDirectory(final String pathString) throws IOException {
```
Yes. After the rename and rewrite I agree. Will move it there.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/utils/SequenceDictionaryUtils.java:
```diff
@@ -466,6 +466,11 @@ private static boolean commonContigsAreAtSameIndices( final Set<String> commonCo
         return contigNames;
     }
 
+    public static List<String> getContigNamesList(final SAMSequenceDictionary refSeqDict) {
```
done

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java:
```diff
@@ -482,4 +482,11 @@ public static int countRightClippedBases(final Cigar cigar) {
         }
     }
+
+    public static int countAlignedBases(final Cigar cigar ) {
```
Add unit test.

vruano replied:
Test Cigar is not Null with ```Utils.nonNull```

SHuang-Broad replied:
done and done

magicDGS commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVFastqUtilsUnitTest.java:
```diff
@@ -29,7 +30,7 @@
  * Unit tests for components in {@link SVFastqUtils}.
  */
-public class SVFastqUtilsUnitTest {
+public class SVFastqUtilsUnitTest extends BaseTest {
```
After https://github.com/broadinstitute/gatk/pull/3475, this should extend `GATKBaseTest`

magicDGS commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVUtilsUnitTest.java:
```diff
@@ -4,12 +4,9 @@
-/**
- * Unit tests for SVUtils.
- */
 public class SVUtilsUnitTest extends BaseTest {
```
After https://github.com/broadinstitute/gatk/pull/3475, this should extend `GATKBaseTest`

SHuang-Broad commented:
@vruano I've addressed more questions. 
@magicDGS sorry but I don't think I have time for the `getRefRDD()` method this or next week. But I've added a todo for it, and will get back on it later after my complex SV work is submitted.

magicDGS commented:
@SHuang-Broad - thanks for taking it into consideration, even if cannot be soon.
